synapse/storage/schema/main/delta/78/03event_extremities_constraints.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2023 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+This migration adds foreign key constraint to `event_forward_extremities` table.
+"""
+from synapse.storage.background_updates import (
+    ForeignKeyConstraint,
+    run_validate_constraint_and_delete_rows_schema_delta,
+)
+from synapse.storage.database import LoggingTransaction
+from synapse.storage.engines import BaseDatabaseEngine
+
+FORWARD_EXTREMITIES_TABLE_SCHEMA = """
+    CREATE TABLE event_forward_extremities2(
+        event_id TEXT NOT NULL,
+        room_id TEXT NOT NULL,
+        UNIQUE (event_id, room_id),
+        CONSTRAINT event_forward_extremities_event_id FOREIGN KEY (event_id) REFERENCES events (event_id)
+    );
+"""
+
+FORWARD_EXTREMITIES_INDICES_SCHEMA = """
+    CREATE INDEX ev_extrem_room ON event_forward_extremities(room_id);
+    CREATE INDEX ev_extrem_id ON event_forward_extremities(event_id);
+"""
```

Comment:
Is there a more automatic way we can copy things instead of relying on our copy-paste skills here to make sure it's right?

Reply:
Actually, probably using the special `sqlite_schema` table. Let me give it a go

Reply:
I have changed it to automatically port the indexes for SQLite, but I can't see a nice way of automating the table creation without having to parse the SQL.

Reply:
Fancy!!

---

From some Googling as well, doesn't seem possible for SQLite out of box :+1: 

With Postgres, we could use `CREATE TABLE new_table LIKE old_table;` ([reference](https://www.postgresql.org/docs/current/sql-createtable.html)) which is exactly what we want to copy the schema/structure only.

SQLite does have [`CREATE TABLE ... AS SELECT`](https://www.sqlite.org/lang_createtable.html#create_table_as_select_statements) but it doesn't copy the primary key, constraints, or defaults. It doesn't copy constraints (I assume that means indexes as well) but we could work around it with the code you already have to copy those over. Perhaps we could overcome all of these problems by applying all indexes/constraints (currently the built-in ones are excluded with `AND sql IS NOT NULL`)

https://github.com/matrix-org/synapse/blob/2451d8d61e40109dff11cbe50cf8729c1cc67ad5/synapse/storage/background_updates.py#L1103-L1110